[media] Enable IAMF MP4 parsing

osagie98 opened the pull request:
b/271301103

osagie98 committed:
[media] Enable IAMF MP4 parsing

b/271301103

osagie98 committed:
Remove references to iamd

osagie98 committed:
Parse and store Config OBUs

osagie98 committed:
Parse config OBUs more efficiently

osagie98 committed:
Update starboard utils

osagie98 committed:
Remove unused iamf box variables

osagie98 committed:
Minor formatting

xiaomings commented on third_party/chromium/media/formats/mp4/box_definitions.h:
```diff
@@ -378,6 +378,44 @@ struct MEDIA_EXPORT OpusSpecificBox : Box {
+  enum IamfObuType {
+    // The following 3 OBUs must be placed in the IA Sequence in the following
+    // order. https://aomediacodec.github.io/iamf/#standalone-descriptor-obus
+    kIAMFObuTypeCodecConfig = 0,
```
nit: kIAMF => kIamf, to be consistent with the type name.

xiaomings commented on third_party/chromium/media/formats/mp4/box_definitions.cc:
```diff
@@ -1438,6 +1438,120 @@ bool OpusSpecificBox::Parse(BoxReader* reader) {
+    }
+    // We don't handle Data OBUs here.
+    default: {
+      RCHECK(static_cast<int>(obu_type) < kIAMFObuTypeMax);
```
This code fails when |obu_type| is |kIAMFObuTypeSequenceHeader|.

Also, since obu types are not continuous, avoid checking using it's max value in general.  Consider listing them separately without a default statement, so compiler can generate warnings whenever there is a mismatch.

xiaomings commented on third_party/chromium/media/formats/mp4/box_definitions.h:
```diff
@@ -378,6 +378,44 @@ struct MEDIA_EXPORT OpusSpecificBox : Box {
+
+  std::vector<uint8_t> config_obus;
+  // Points to the next location in |config_obus| to be read into.
+  int buffer_pos = 0;
```
It's ok to keep this here for now.  Consider removing it and refactoring the code as this member variable shouldn't be part of the public interface of this structure.

xiaomings commented on third_party/chromium/media/formats/mp4/box_definitions.h:
```diff
@@ -378,6 +378,44 @@ struct MEDIA_EXPORT OpusSpecificBox : Box {
+struct MEDIA_EXPORT IamfSpecificBox : Box {
+  DECLARE_BOX_METHODS(IamfSpecificBox);
+
+  enum IamfObuType {
```
nit: type declaration before functions (i.e. DECLARE_BOX_METHODS above), and functions (e.g. `ReadSequenceHeader()`) before member variables.

osagie98 committed:
Formatting and optimization

osagie98 committed:
Further formatting

osagie98 committed:
Added parsing CHECKs

osagie98 committed:
Remove unnecessary variable

osagie98 committed:
Set decoder config extra_data

xiaomings commented on third_party/chromium/media/formats/mp4/mp4_stream_parser.cc:
```diff
@@ -882,6 +910,17 @@ ParseResult MP4StreamParser::EnqueueSample(BufferQueueMap* buffers) {
       return ParseResult::kError;
 #endif  // BUILDFLAG(USE_PROPRIETARY_CODECS)
     }
+#if defined(STARBOARD)
```
There was a section in the previous code (http://shortn/_P2VQDJNMcR) with incorrect indent, now it's overwritten and the indent is good.

xiaomings commented on third_party/chromium/media/formats/mp4/mp4_stream_parser.cc:
```diff
@@ -385,6 +388,14 @@ bool MP4StreamParser::ParseMoov(BoxReader* reader) {
+      } else if (audio_format == FOURCC_IAMF) {
+        codec = AudioCodec::kIAMF;
+        profile = entry.iamf.primary_profile == 0 ? AudioCodecProfile::kIAMF_SIMPLE : AudioCodecProfile::kIAMF_BASE;
+        // Default values.
```
In that case, let's add some comments here mentioning that.

osagie98 commented on third_party/chromium/media/base/audio_codecs.h:
```diff
@@ -33,13 +33,20 @@ enum class AudioCodec {
   kAC3 = 16,
   kMpegHAudio = 17,
+#if defined (STARBOARD)
+  kIAMF = 18,
```
Good catch, I'll see if I can check with Chromium

xiaomings replied:
Chromium requires the enum values to be fixed to avoid affect telemetry info.  It's great if we can align with them, but if we cannot find out the exact value they are going to use, it's ok to use a tentative value, as long as this doesn't pollute our telemetry, where Thor should have more info.

osagie98 committed:
Correct indentation

osagie98 committed:
Add comment explaining default IAMF config values

osagie98 commented on third_party/chromium/media/formats/mp4/box_reader.h:
```diff
@@ -81,6 +81,31 @@ class MEDIA_EXPORT BufferReader {
+    for (int i = 0; i < max_encoded_bytes; ++i) {
+      uint8_t* byte = &buffer[i];
+      RCHECK(Read1(byte));
+      *encoded_value |= ((*byte & 0x7f) << (i * 7));
```
I see what you're saying. Is it possible the rvalue returned by `((*byte & 0x7f) << (i * 7))` is only a byte wide? I assumed it would resolve to an int large enough to hold the value.

osagie98 committed:
Update leb128 parsing

osagie98 committed:
Add Config OBU parsing TODO

osagie98 committed:
Read Config OBUs with BufferReader

osagie98 committed:
Add CHECK for BufferReader

osagie98 commented on third_party/chromium/media/formats/mp4/box_definitions.cc:
```diff
@@ -1438,6 +1438,101 @@ bool OpusSpecificBox::Parse(BoxReader* reader) {
+FourCC IamfSpecificBox::BoxType() const { return FOURCC_IAMF; }
+
+bool IamfSpecificBox::Parse(BoxReader* reader) {
+  config_obus.resize(reader->box_size() - reader->pos());
```
Done

xiaomings commented on third_party/chromium/media/formats/mp4/box_definitions.h:
```diff
@@ -378,6 +378,31 @@ struct MEDIA_EXPORT OpusSpecificBox : Box {
+  bool ReadOBU(BufferReader* reader);
+  bool ReadOBUHeader(BufferReader* reader, uint8_t* obu_type,
+                     uint32_t* obu_size);
+  bool ReadLeb128Value(BufferReader* reader, uint32_t* encoded_value) const ;
```
`encoded_value` => `value`, as `encoded_value` can be misleading.  Also remove the redundant space between `const` and ';'.

xiaomings commented on third_party/chromium/media/formats/mp4/box_definitions.cc:
```diff
@@ -1438,6 +1438,113 @@ bool OpusSpecificBox::Parse(BoxReader* reader) {
+      RCHECK(reader->Read1(&profile));
+      RCHECK(profile <= 1);
+
+      RCHECK(reader->SkipBytes(sizeof(uint8_t)));
```
Redundant, as all remaining bytes will be skipped ein line 1497.

xiaomings commented on third_party/chromium/media/formats/mp4/box_definitions.cc:
```diff
@@ -1438,6 +1438,113 @@ bool OpusSpecificBox::Parse(BoxReader* reader) {
+  uint8_t header_flags;
+  RCHECK(reader->Read1(&header_flags));
+  *obu_type = (header_flags >> 3) & 0x1f;
+  bool obu_redundant_copy = (header_flags >> 2) & 1;
+  bool obu_trimming_status_flag = (header_flags >> 1) & 1;
+  bool obu_extension_flag = header_flags & 1;
```
nit: empty line before and after.

xiaomings commented on third_party/chromium/media/formats/mp4/box_definitions.cc:
```diff
@@ -1438,6 +1438,113 @@ bool OpusSpecificBox::Parse(BoxReader* reader) {
+// https://aomediacodec.github.io/av1-spec/#leb128.
+bool IamfSpecificBox::ReadLeb128Value(BufferReader* reader,
+                                      uint32_t* encoded_value) const {
+  CHECK(reader);
+  CHECK(encoded_value);
```
DCHECK is good enough here.  CHECK() has production cost, and should be reserved for the cases where integrity is important and hard to verify.

osagie98 committed:
Remove redundancy in IamfSpecificBox

Also removes directives checking for STARBOARD in the audio codecs
files, as directives are not used for other codecs.

osagie98 committed:
Improve iamf codec string parsing
